### PR TITLE
61 add version control

### PR DESCRIPTION
**Pull request by gerrycampion**
- Rules are now versioned when Patched, Published or Deleted
- Deleted rules are soft-deleted
- A previous version of the rule can be restored

**Versioning**
Each version of a rule is stored as a separate item in the container.
Versions are maintained using the "created" timestamp property.
Versions are grouped using the "latestId" property, which points to the id of the current version.
If the item is the latest item, it will not have a "latestId" property
If the item has been deleted, the "latestId" property will be empty and if the rule is published, it will be unpublished so that the library/engine do not think the rule exists.
If a published item is patched, the new item will remain published and the previous version will be unpublished.

**Testing**
The following functionality should be tested:
- Getting all rules
- Getting a version of a rule
- Creating a new rule
- Modifying a rule
- Restoring a previous version of a rule
- Deleting a rule
- Publishing a rule

### CODE REVIEW

File: api/providers/BaseUsers.ts [hunk 1]
```diff
@@ -9,3 +11,49 @@ const USERS_PROVIDERS: { [provider: string]: IUsers } = {
...
+  ]);
+  for (const rule of rules) {
+    rule.creator = users[rule.creator.id] ?? {
+      id: rule.creator.id,
+      name: null,
+    };
+  }
+}
+
+export async function addUsernamesToRule(rule: IRule) {
```
Comment (nhaydel) [inline api/providers/BaseUsers.ts:47]: Why is this needed?

Comment (gerrycampion) [reply api/providers/BaseUsers.ts:47]: It allows you to see the usernames (along with the timestamps) in the version picker: https://github.com/cdisc-org/conformance-rules-editor/blob/fe3154b77c0736e7d155d5e4358bc9d392562731/src/components/YamlEditor/DiffEditor.tsx#L77

**Comment by nhaydel** [comment]
@gerrycampion overall the code looks good. I am a little concerned about saving a copy of the rule in the rules table on each save. I think if enough users save frequently the table could get pretty large and queries could slow down. What do you think about creating a new table: RuleHistory, which would store old versions of the rule? The rule record in the main rules table would have references to the rule history table:
```
[
{
    "created": timestamp
    "creator": <name>
    "replaced/updated": <timestamp when a new version of the rule became the latest>
    "rule_history_id": "id of the rule version in the rule history table"
}
]
```

This way we would only query rule history when getting a specific diff, so we could look up by ID when the rule author selects a rule to diff with. It might also be cheaper since we would not be doing as many queries on all rules ever created

**Comment by gerrycampion** [comment]
> @gerrycampion overall the code looks good. I am a little concerned about saving a copy of the rule in the rules table on each save. I think if enough users save frequently the table could get pretty large and queries could slow down. What do you think about creating a new table: RuleHistory, which would store old versions of the rule? The rule record in the main rules table would have references to the rule history table:
> 
> ```
> [
> {
>     "created": timestamp
>     "creator": <name>
>     "replaced/updated": <timestamp when a new version of the rule became the latest>
>     "rule_history_id": "id of the rule version in the rule history table"
> }
> ]
> ```
> 
> This way we would only query rule history when getting a specific diff, so we could look up by ID when the rule author selects a rule to diff with. It might also be cheaper since we would not be doing as many queries on all rules ever created

The Rules table now stores a historical list of `{ created, creator, id}`. The historical rules are stored in a separate container.
I also updated the core id population to avoid reusing core ids for rules that have been deleted.

**Review by nhaydel — APPROVED** [review]
